### PR TITLE
Spring Boot Portfolio Project — Part 15. - ShoppingCart tests + added fixes from last PR review

### PR DESCRIPTION
**Pull request by dedis34**

### CODE REVIEW

File: src/test/java/org/example/controller/ShoppingCartControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import java.util.HashSet;
+import java.util.Set;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.*;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+class ShoppingCartControllerTest {
```
Comment (slade13) [inline src/test/java/org/example/controller/ShoppingCartControllerTest.java:25]: You can add edge cases for your tests like empty cart or non existing user. Add test for item not found, what if updateCartItem throws exception?

File: pom.xml
Comment (slade13) [inline pom.xml:1]: You can add here jacoco plugin to track test coverage.

File: src/test/java/org/example/controller/ShoppingCartControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    void updateCartItem_WhenCartItemExists_ShouldUpdateQuantity() throws Exception {
+        UpdateCartItemRequestDto request = new UpdateCartItemRequestDto(3);
+        when(shoppingCartServiceImpl.updateCartItem(eq(1L), any(UpdateCartItemRequestDto.class)))
+                .thenReturn(shoppingCartResponseDto);
+
+        mockMvc.perform(put("/api/cart/cart-items/1")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content("{\"quantity\":3}"))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.id").value(1L));
```
Comment (slade13) [inline src/test/java/org/example/controller/ShoppingCartControllerTest.java:63]: Use ObjectMapper instead of raw JSON string.

File: src/test/java/org/example/controller/ShoppingCartControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        Set<CartItemResponseDto> cartItems = new HashSet<>();
+        cartItems.add(new CartItemResponseDto(1L, 1L, "Sample Book", 2));
+        shoppingCartResponseDto = new ShoppingCartResponseDto(1L, 1L, cartItems);
+
+        mockMvc = MockMvcBuilders.standaloneSetup(shoppingCartController).build();
+    }
+
+    @Test
+    @WithMockUser(roles = "USER")
```
Comment (slade13) [inline src/test/java/org/example/controller/ShoppingCartControllerTest.java:53]: Test also for unauthorized access - verify 403/401 if no user/auth.

File: src/test/java/org/example/controller/ShoppingCartControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    void updateCartItem_WhenCartItemExists_ShouldUpdateQuantity() throws Exception {
+        UpdateCartItemRequestDto request = new UpdateCartItemRequestDto(3);
```
Comment (slade13) [inline src/test/java/org/example/controller/ShoppingCartControllerTest.java:55]: You can test invalid input e.g., { "quantity": -1 } should return 400.

File: src/test/java/org/example/service/ShoppingCartServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    void addToCart_WhenUserExistsAndBookExists_ShouldAddItemToCart() {
+        AddItemToCartRequestDto request = new AddItemToCartRequestDto(1L, 2);
+        when(userRepository.findById(1L)).thenReturn(Optional.of(user));
+        when(shoppingCartRepository.findByUser(user)).thenReturn(Optional.of(shoppingCart));
+        when(bookRepository.findById(1L)).thenReturn(Optional.of(book));
+        when(shoppingCartMapper.toResponseDto(any(ShoppingCart.class))).thenReturn(shoppingCartResponseDto);
+
+        ShoppingCartResponseDto result = shoppingCartService.addToCart(1L, request);
+
+        assertNotNull(result);
```
Comment (slade13) [inline src/test/java/org/example/service/ShoppingCartServiceTest.java:90]: Instead of `assertNotNull(result)`, check actual fields in `result`.

File: src/test/java/org/example/service/ShoppingCartServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        cartItem.setBook(book);
+        cartItem.setQuantity(2);
+        cartItem.setShoppingCart(shoppingCart);
+
+        shoppingCartResponseDto = new ShoppingCartResponseDto(1L, 1L, new HashSet<>());
+    }
+
+    @Test
+    void addToCart_WhenUserExistsAndBookExists_ShouldAddItemToCart() {
+        AddItemToCartRequestDto request = new AddItemToCartRequestDto(1L, 2);
```
Comment (slade13) [inline src/test/java/org/example/service/ShoppingCartServiceTest.java:82]: Improve DTO readability by using builders or test factories for DTOs.

File: src/test/java/org/example/service/ShoppingCartServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        cartItem.setId(1L);
+        cartItem.setBook(book);
+        cartItem.setQuantity(2);
+        cartItem.setShoppingCart(shoppingCart);
+
+        shoppingCartResponseDto = new ShoppingCartResponseDto(1L, 1L, new HashSet<>());
+    }
+
+    @Test
+    void addToCart_WhenUserExistsAndBookExists_ShouldAddItemToCart() {
```
Comment (slade13) [inline src/test/java/org/example/service/ShoppingCartServiceTest.java:81]: No test for book not found, add a case where book repo returns Optional.empty().

**Review by slade13 — APPROVED** [review]
Great job! Please review my comments. :)
